### PR TITLE
ci: Merge similar try jobs when possible

### PR DESCRIPTION
**Pull request by mrobinson**
This comes up a lot when triggering wpt-2013 and wpt-2020. Instead of
merging the jobs and triggering both layouts, the try parser will
trigger two separate jobs. Running two of the same builds at once seems
to cause the CI to fail one of them [1]. This fixes that issue.

1. An example of this: https://github.com/servo/servo/actions/runs/7892269495

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] There are tests for these changes

### CODE REVIEW

**Comment by sagudev** [comment]
When triggering WPT for both layouts there is `wpt` preset, so just creating `T-wpt` label should run wpt on both layouts. That being said is still think it's good we do merging if possible.

The error you se are due to how artifacts are shared, that I hope will be solved with artifacts-v4 that I am working on.

File: python/servo/try_parser.py [hunk 1]
```diff
@@ -68,15 +80,15 @@ def handle_preset(s: str) -> Optional[JobConfig]:
     elif s in ["wpt", "linux-wpt"]:
         return JobConfig("Linux WPT", Workflow.LINUX, unit_tests=True, wpt_layout=Layout.all())
     elif s in ["wpt-2013", "linux-wpt-2013"]:
-        return JobConfig("Linux WPT legacy-layout", Workflow.LINUX, wpt_layout=Layout.layout2013)
+        return JobConfig("Linux WPT", Workflow.LINUX, wpt_layout=Layout.layout2013)
```
Comment (sagudev) [inline python/servo/try_parser.py:83]: Can we just keep different names here and ignore name field when merging.

Comment (mrobinson) [reply python/servo/try_parser.py:83]: The problem is that the merged job will keep the name of the first job, which will be incorrect. I would like to make a later change that generates the job name lazily, but for now this is the simplest approach.

File: python/servo/try_parser.py [hunk 2]
```diff
@@ -55,6 +57,16 @@ class JobConfig(object):
     unit_tests: bool = False
     wpt_tests_to_run: str = ""
 
+    def merge(self, other: JobConfig) -> bool:
+        """Try to merge another job with this job. Returns True if merging is successful
+           or False if not. If merging is successful this job will be modified."""
+        if self.workflow != other.workflow or self.profile != other.profile or \
+                self.wpt_tests_to_run != other.wpt_tests_to_run:
+            return False
```
[the remark this answers is not in the document]
Comment (mrobinson) [reply python/servo/try_parser.py]: That's an interesting idea, but it seems a bit heavy-weight and people adding new fields to JobConfig will still have to remember to update this function. Instead I've created a new class variable which holds the fields that cannot differ for a merge to happen. This is near the fields and I've added doc to remind people to update the list.

In addition, I've added more tests, which should also help prevent mistakes in the future.

File: python/servo/try_parser.py (hunk 1 above)
Comment (sagudev) [reply python/servo/try_parser.py:87]: Sounds good to me!

File: python/servo/try_parser.py [hunk 3]
```diff
@@ -192,6 +214,43 @@ def test_empty(self):
...
+                                  'wpt_layout': 'all',
+                                  'wpt_tests_to_run': ''
+                              }]
+                              })
+
+        # These two jobs should not be merged to one that turns on unit tests.
+        a = JobConfig("Linux", Workflow.LINUX, unit_tests=True)
+        b = JobConfig("Linux", Workflow.LINUX, unit_tests=False)
+        self.assertTrue(a.merge(b))
+        self.assertEqual(a, JobConfig("Linux", Workflow.LINUX, unit_tests=True))
```
Comment (sagudev) [inline python/servo/try_parser.py:234]: I think the comment is wrong here as jobs are merged to one, fight?

File: python/servo/try_parser.py [hunk 4]
```diff
@@ -192,6 +214,43 @@ def test_empty(self):
...
+
+        # These two jobs should not be mergable to one that turns on unit tests.
+        a = JobConfig("Linux", Workflow.LINUX, unit_tests=True)
+        b = JobConfig("Linux", Workflow.MACOS, unit_tests=True)
+        self.assertFalse(a.merge(b))
+        self.assertEqual(a, JobConfig("Linux", Workflow.LINUX, unit_tests=True))
```
Comment (sagudev) [inline python/servo/try_parser.py:240]: This comment is wrong to as it compares different workflows.

File: python/servo/try_parser.py [hunk 5]
```diff
@@ -192,6 +214,43 @@ def test_empty(self):
...
+
+        # These two jobs should not be merged to one that turns on unit tests.
+        a = JobConfig("Linux", Workflow.LINUX, unit_tests=True)
+        b = JobConfig("Linux", Workflow.LINUX, unit_tests=False)
+        self.assertTrue(a.merge(b))
+        self.assertEqual(a, JobConfig("Linux", Workflow.LINUX, unit_tests=True))
+
+        # These two jobs should not be mergable to one that turns on unit tests.
+        a = JobConfig("Linux", Workflow.LINUX, unit_tests=True)
+        b = JobConfig("Linux", Workflow.MACOS, unit_tests=True)
```
Comment (sagudev) [inline python/servo/try_parser.py:238]:
```suggestion
        b = JobConfig("Mac", Workflow.MACOS, unit_tests=True)
```

File: python/servo/try_parser.py [hunk 6]
```diff
@@ -192,6 +214,43 @@ def test_empty(self):
...
+        self.assertFalse(a.merge(b))
+        self.assertEqual(a, JobConfig("Linux", Workflow.LINUX, unit_tests=True))
+
+        # Ditto.
```
Comment (sagudev) [inline python/servo/try_parser.py:242]: No comment is better then ditto comment IMO.

File: python/servo/try_parser.py (hunk 3 above)
Comment (mrobinson) [reply python/servo/try_parser.py]: I think my comment was written in a pretty confusing way. I've removed all these comments and converted them to assert messages, which is more useful, I think.

File: python/servo/try_parser.py [hunk 7]
```diff
@@ -192,6 +214,40 @@ def test_empty(self):
...
+        b = JobConfig("Mac", Workflow.MACOS, unit_tests=True)
+        self.assertFalse(a.merge(b), "Should not merge jobs with different workflows.")
+        self.assertEqual(a, JobConfig("Linux", Workflow.LINUX, unit_tests=True))
+
+        a = JobConfig("Linux", Workflow.LINUX, unit_tests=True)
+        b = JobConfig("Linux", Workflow.LINUX, unit_tests=True, profile="production")
+        self.assertFalse(a.merge(b), "Should not merge jobs with different profiles.")
+        self.assertEqual(a, JobConfig("Linux", Workflow.LINUX, unit_tests=True))
+
+        # Ditto.
```
Comment (sagudev) [inline python/servo/try_parser.py:245]:
```suggestion
```

**Review by sagudev — APPROVED** [review]
One suggestion remaining but otherwise looks good.

**Review by mrobinson — COMMENTED** [review]
Thanks for the review, @sagudev.

**Pull request merged**